### PR TITLE
OAuth integration

### PR DESCRIPTION
**Pull request by see-quick**
This PR adds support for OAuth inside the test container. Moreover, now one can use two authentication types:
1. OAuth over plain 
2. OAuth with token bearer
Those types are automatically configured on the server side (e.g., Kafka broker). So, if someone wants to test it, they only need to configure clients.

Also, I have added a dependency `testcontainers-keycloak`, which greatly simplifies the testing process. This should solve [1] together with [2].

Update: It also adds a lot of UTs and ITs.

[1] - https://github.com/strimzi/test-container/issues/63
[2] - https://github.com/strimzi/test-container-images/pull/29

### CODE REVIEW

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 1]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
         return self();
     }
 
+    /**
+     * Fluent method to configure OAuth settings.
+     *
+     * @param keycloakRealm         the Keycloak realm
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:742]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 2]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
...
+     * @param clientId              the OAuth client ID
+     * @param clientSecret          the OAuth client secret
+     * @param keycloakOAuthUri      the Keycloak OAuth URI
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:745]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 3]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
...
+     * @param oauthPreferredUsername the preferred username claim for OAuth
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:746]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 4]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
...
+     * @return {@code StrimziKafkaContainer} instance
+     */
+    public StrimziKafkaContainer withOAuthConfig(final String keycloakRealm,
+                                                 final String clientId,
+                                                 final String clientSecret,
+                                                 final String keycloakOAuthUri,
+                                                 final String oauthPreferredUsername,
+                                                 final List<String> superUsers) {
+        this.oauthEnabled = true;
+        this.keycloakRealm = keycloakRealm;
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:757]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 5]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
...
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.keycloakOauthUri = keycloakOAuthUri;
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:760]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 6]
```diff
@@ -566,9 +736,112 @@ public StrimziKafkaContainer withProxyContainer(final ToxiproxyContainer proxyCo
...
+        this.oauthPreferredUsername = oauthPreferredUsername;
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:761]: Same comment as previously.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 7]
```diff
@@ -88,6 +92,21 @@ public class StrimziKafkaContainer extends GenericContainer<StrimziKafkaContaine
     private ToxiproxyClient toxiproxyClient;
     private Proxy proxy;
 
+    // OAuth fields
+    private boolean oauthEnabled;
+    private String keycloakRealm;
+    private String clientId;
+    private String clientSecret;
+    private String keycloakOauthUri;
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:100]: Same comment as for `keycloakRealm`.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 8]
```diff
@@ -396,21 +445,131 @@ private Properties buildDefaultServerProperties() {
...
+     * @param properties The Kafka server properties.
+     */
+    private void setSuperUsersIntoProperties(Properties properties) {
+        if (this.superUsers != null && !this.superUsers.isEmpty()) {
+            String superUsersProperty = this.superUsers.stream()
+                .map(user -> "User:" + user)
+                .collect(Collectors.joining(";"));
+            properties.setProperty("super.users", superUsersProperty);
+        }
+    }
```
Comment (mstruk) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:565]: You have the ability to set `super.users` but I don't see in any configuration that you specify the `authorizer.class.name`. Without specifying that in the Kafka broker configuration there is no active authorization in place, and setting `super.users` becomes irrelevant (ignored).

**Review by mstruk — COMMENTED** [review]
Some initial comments ...

File: src/main/java/io/strimzi/test/container/AuthenticationType.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * OAuth Bearer token authentication.
+     */
+    OAUTH_BEARER,
+
+    /**
+     * Mutual TLS authentication.
+     */
+    MUTUAL_TLS,
```
[the remark this answers is not in the document]
Comment (see-quick) [reply src/main/java/io/strimzi/test/container/AuthenticationType.java:25]: Sure, I can remove it from the list.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 10]
```diff
@@ -88,6 +92,21 @@ public class StrimziKafkaContainer extends GenericContainer<StrimziKafkaContaine
     private ToxiproxyClient toxiproxyClient;
     private Proxy proxy;
 
+    // OAuth fields
+    private boolean oauthEnabled;
+    private String keycloakRealm;
```
[the remark this answers is not in the document]
Comment (see-quick) [reply src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:97]: > I depends on how modular / pluggable / extensible / generic you want your code to be here

The goal for now is to stick with just Keycloak. And possibly in the future we can add more support for other AS (f.e., hydra).

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 11]
```diff
@@ -354,15 +375,26 @@ private String randomUuid() {
...
         // Default properties for server.properties
         Properties properties = new Properties();
 
         // Common settings for both KRaft and non-KRaft modes
-        properties.setProperty("listeners", "PLAINTEXT://:9092");
-        properties.setProperty("inter.broker.listener.name", "PLAINTEXT");
-        properties.setProperty("advertised.listeners", "PLAINTEXT://localhost:9092");
-        properties.setProperty("listener.security.protocol.map", "PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL");
+        properties.setProperty("listeners", listeners);
+        properties.setProperty("inter.broker.listener.name", "BROKER1");
```
[the remark this answers is not in the document]
Comment (see-quick) [reply src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:394]: True. I have updated the configs. Now it looks a little bit better and it's dynamic so a lot of code is simpler.

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 12]
```diff
@@ -396,21 +445,131 @@ private Properties buildDefaultServerProperties() {
...
+            "org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required " +
+                "oauth.client.id=\"%s\" " +
+                "oauth.client.secret=\"%s\" " +
+                "oauth.token.endpoint.uri=\"%s\" " +
+                "oauth.username.claim=\"%s\";",
+            this.clientId,
+            this.clientSecret,
+            this.keycloakOauthUri + "/realms/" + this.keycloakRealm + "/protocol/openid-connect/token",
+            this.oauthPreferredUsername // e.g., "preferred_username"
+        );
```
[the remark this answers is not in the document]
Comment (see-quick) [reply src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:507]: > oauth.jwks.endpoint.uri, and oauth.valid.issuer.uri in addition to what you have specified.

Those are set via ENV vars i.e.,
```java
if (this.isOAuthEnabled()) {
            // Set OAuth environment variables (using properties does not propagate to System properties)
            this.addEnv("OAUTH_JWKS_ENDPOINT_URI", this.keycloakOauthUri + "/realms/" + this.keycloakRealm + "/protocol/openid-connect/certs");
            this.addEnv("OAUTH_VALID_ISSUER_URI", this.keycloakOauthUri + "/realms/" + this.keycloakRealm);
            this.addEnv("OAUTH_CLIENT_ID", this.clientId);
            this.addEnv("OAUTH_CLIENT_SECRET", this.clientSecret);
            this.addEnv("OAUTH_TOKEN_ENDPOINT_URI", this.keycloakOauthUri + "/realms/" + this.keycloakRealm + "/protocol/openid-connect/token");
            this.addEnv("OAUTH_USERNAME_CLAIM", this.usernameClaim);
        }
 ```

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 13]
```diff
@@ -396,21 +445,131 @@ private Properties buildDefaultServerProperties() {
...
+
+        // Dynamically build the 'super.users' property
+        this.setSuperUsersIntoProperties(properties);
+
+        // Construct the JAAS configuration with configurable username and password
+        final String jaasConfig = String.format(
+            "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"%s\" password=\"%s\";",
+            this.saslUsername,
+            this.saslPassword
+        );
```
[the remark this answers is not in the document]
Comment (see-quick) [reply src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:468]: Yes I have tested both `JaasServerOauthOverPlainValidatorCallbackHandler` and `JaasServerOauthValidatorCallbackHandler`

**Review by mstruk — APPROVED** [review]
For the review I needed to see logging from Strimzi OAuth library, and the effective `server.properties` file. I was able to achieve that by editing the `StrimziKafkaContainer.java` code:
```
diff --git a/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java b/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java
index 010643a..704dd3a 100644
--- a/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java
+++ b/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java
@@ -269,7 +269,11 @@ public class StrimziKafkaContainer extends GenericContainer<StrimziKafkaContaine
             }
 
             command += "bin/kafka-storage.sh format -t=\"" + this.clusterId + "\" -c /opt/kafka/config/kraft/server.properties \n";
-            command += "bin/kafka-server-start.sh /opt/kafka/config/kraft/server.properties \n";
+            command += "cp /opt/kafka/config/log4j.properties /tmp/log4j.properties\n";
+            command += "echo \"log4j.logger.io.strimzi=DEBUG\" >> /tmp/log4j.properties\n";
+            command += "cat /opt/kafka/config/kraft/server.properties\n";
+            command += "export KAFKA_LOG4J_OPTS=\"-Dlog4j.configuration=file:/tmp/log4j.properties\"\n";
+            command += "bin/kafka-server-start.sh /opt/kafka/config/kraft/server.properties\n";
         }
 
         Utils.asTransferableBytes(serverPropertiesFile).ifPresent(properties -> copyFileToContainer(
```
It would be nice to have a way to configure logging and turning on the dumping to the console of `server.properties`, but that may be addressed in another PR.

Also a few words in the `README.md` about how to troubleshoot the problems with Podman timeouts would be very welcome. For me it helped to run:

```
podman server ssh --username root
echo '[engine]
service_timeout=0' >> /etc/containers/containers.conf
```
Although, I still see some timeouts and tests stalling, which seems to happen due to resource exhaustion as a consequence of containers spinned up for tests not being cleaned up as they should.

File: src/test/java/io/strimzi/test/container/StrimziKafkaContainerTest.java [hunk 14]
```diff
@@ -4,26 +4,712 @@
...
+            "org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required " +
+                "oauth.client.id=\"%s\" " +
+                "oauth.client.secret=\"%s\" " +
+                "oauth.token.endpoint.uri=\"%s\" " +
+                "oauth.username.claim=\"%s\";",
+            kafkaContainer.getClientId(),
+            kafkaContainer.getClientSecret(),
+            kafkaContainer.getOauthUri() + "/realms/" + kafkaContainer.getRealm() + "/protocol/openid-connect/token",
+            kafkaContainer.getUsernameClaim()
+        );
```
Comment (mstruk) [inline src/test/java/io/strimzi/test/container/StrimziKafkaContainerTest.java:400]:
```suggestion
        String expectedJaasConfig = "org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ";
```

**Comment by see-quick** [comment]
> It would be nice to have a way to configure logging and turning on the dumping to the console of `server.properties`, but that may be addressed in another PR.

I think this method already exists but needs a few modifications to also change `io.strimzi` logger
```java
 public StrimziKafkaContainer withKafkaLog(Level level) {
        String log4jConfig = "log4j.rootLogger=" + level.name() + ", stdout\n" +
            "log4j.appender.stdout=org.apache.log4j.ConsoleAppender\n" +
            "log4j.appender.stdout.layout=org.apache.log4j.PatternLayout\n" +
            "log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c)%n\n";

        // Copy the custom log4j.properties into the container
        this.withCopyToContainer(
            Transferable.of(log4jConfig.getBytes(StandardCharsets.UTF_8)),
            "/opt/kafka/config/log4j.properties"
        );

        return self();
    }
```

> Also a few words in the README.md about how to troubleshoot the problems with Podman timeouts would be very welcome. 

Yeah I agree I have opened two issues to cover those problems. Thanks for the review it helped a lot! @mstruk

**Pull request merged**